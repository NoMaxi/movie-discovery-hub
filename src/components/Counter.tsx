import React from "react";

interface CounterProps {
    initialValue?: number;
}

interface CounterState {
    count: number;
}

class Counter extends React.Component<CounterProps, CounterState> {
    constructor(props: CounterProps) {
        super(props);
        this.state = {
            count: props.initialValue || 0,
        };
        this.increment = this.increment.bind(this);
        this.decrement = this.decrement.bind(this);
    }

    increment() {
        this.setState((prevState: CounterState) => ({
            count: prevState.count + 1,
        }));
    }

    decrement() {
        this.setState((prevState: CounterState) => ({
            count: prevState.count - 1,
        }));
    }

    render() {
        const isDecrementDisabled = this.state.count <= 0;

        return React.createElement(
            "div",
            { className: "bg-gray-800 p-4 flex items-center justify-center gap-4" },
            React.createElement(
                "button",
                {
                    className: `bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded-full cursor-pointer ${ isDecrementDisabled ? "opacity-50 cursor-not-allowed" : ""
                    }`,
                    onClick: this.decrement,
                    disabled: isDecrementDisabled,
                },
                "-"
            ),
            React.createElement("span", { className: "text-white text-xl font-bold" }, this.state.count),
            React.createElement(
                "button",
                {
                    className: "bg-red-500 hover:bg-red-700 text-white font-bold py-2 px-4 rounded-full cursor-pointer",
                    onClick: this.increment,
                },
                "+"
            )
        );
    }
}

export default Counter;
