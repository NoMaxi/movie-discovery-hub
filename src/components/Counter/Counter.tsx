import React from "react";

interface CounterProps {
    initialValue?: number;
}

interface CounterState {
    count: number;
}

class Counter extends React.Component<CounterProps, CounterState> {
    state: CounterState = {
        count: this.props.initialValue ?? 0,
    };

    increment = () => {
        this.setState((prevState: CounterState) => ({
            count: prevState.count + 1,
        }));
    };

    decrement = () => {
        this.setState((prevState: CounterState) => ({
            count: prevState.count - 1,
        }));
    };

    render() {
        const isDecrementDisabled = this.state.count <= 0;

        return React.createElement(
            "div",
            { className: "p-4 flex items-center justify-center gap-4" },
            React.createElement(
                "button",
                {
                    className: `bg-[var(--color-primary)] hover:brightness-80 active:brightness-60 w-[40px] h-[40px] 
                    text-xl font-bold py-2 px-4 rounded-full cursor-pointer
                    ${ isDecrementDisabled ? "opacity-50 cursor-not-allowed" : "" }`,
                    onClick: this.decrement,
                    disabled: isDecrementDisabled,
                },
                "-",
            ),
            React.createElement(
                "div",
                {
                    className:
                        "flex items-center justify-center text-3xl font-bold min-w-[30px]",
                },
                this.state.count,
            ),
            React.createElement(
                "button",
                {
                    className: `bg-[var(--color-primary)] hover:brightness-80 active:brightness-60 w-[40px] h-[40px]
                    text-xl font-bold py-2 px-4 rounded-full cursor-pointer`,
                    onClick: this.increment,
                },
                "+",
            ),
        );
    }
}

export default Counter;
