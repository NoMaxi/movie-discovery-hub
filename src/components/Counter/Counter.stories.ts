import type { Meta, StoryObj } from "@storybook/react";
import Counter from "./Counter";

const meta = {
    component: Counter,
} satisfies Meta<typeof Counter>;

export default meta;

type Story = StoryObj<typeof meta>;

export const WithInitialValue: Story = {
    args: {
        initialValue: 10,
    },
};

export const WithoutInitialValue: Story = {
    args: {},
};
