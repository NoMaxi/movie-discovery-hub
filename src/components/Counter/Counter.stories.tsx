import type { Meta, StoryObj } from "@storybook/react";
import Counter from "./Counter";

const meta = {
    component: Counter,
    argTypes: {
        initialValue: {
            control: "number",
            description: "Initial value for the counter",
        },
    },
} satisfies Meta<typeof Counter>;

export default meta;

type Story = StoryObj<typeof meta>;

export const WithInitialValue: Story = {
    args: {
        initialValue: 10,
    },
};

export const WithoutInitialValue: Story = {
    args: {},
};
