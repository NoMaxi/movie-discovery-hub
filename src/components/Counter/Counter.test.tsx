import { render, fireEvent, screen } from "@testing-library/react";
import "@testing-library/jest-dom";
import { Counter } from "./Counter";

describe("Counter", () => {
    test("Should render the default initial value if not provided in props", () => {
        const { asFragment } = render(<Counter />);

        expect(screen.getByText("0")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should render the initial value provided in props", () => {
        const { asFragment } = render(<Counter initialValue={5} />);

        expect(screen.getByText("5")).toBeInTheDocument();
        expect(asFragment()).toMatchSnapshot();
    });

    test("Should decrement the displayed value on decrement button click", () => {
        render(<Counter initialValue={5} />);
        const decrementButton = screen.getByText("-");

        fireEvent.click(decrementButton);

        expect(screen.getByText("4")).toBeInTheDocument();

        fireEvent.click(decrementButton);

        expect(screen.getByText("3")).toBeInTheDocument();
    });

    test("Should increment the displayed value on increment button click", () => {
        render(<Counter initialValue={5} />);
        const incrementButton = screen.getByText("+");

        fireEvent.click(incrementButton);

        expect(screen.getByText("6")).toBeInTheDocument();

        fireEvent.click(incrementButton);

        expect(screen.getByText("7")).toBeInTheDocument();
    });

    test("Should disable the decrement button if the count is 0", () => {
        render(<Counter initialValue={0} />);
        const decrementButton = screen.getByText("-");

        expect(decrementButton).toBeDisabled();

        expect(screen.getByText("0")).toBeInTheDocument();
    });

    test("Should disable the decrement button if the count is less than 0", () => {
        render(<Counter initialValue={-5} />);
        const decrementButton = screen.getByText("-");

        expect(decrementButton).toBeDisabled();

        expect(screen.getByText("-5")).toBeInTheDocument();
    });

    test("Should disable the decrement button when count reaches 0 after decrementing", () => {
        render(<Counter initialValue={1} />);
        const decrementButton = screen.getByText("-");

        expect(decrementButton).toBeEnabled();

        fireEvent.click(decrementButton);

        expect(screen.getByText("0")).toBeInTheDocument();
        expect(decrementButton).toBeDisabled();
    });

    test("Should re-enable the decrement button and decrement correctly after being disabled", () => {
        render(<Counter initialValue={0} />);
        const decrementButton = screen.getByText("-");
        const incrementButton = screen.getByText("+");

        expect(decrementButton).toBeDisabled();

        fireEvent.click(incrementButton);

        expect(screen.getByText("1")).toBeInTheDocument();
        expect(decrementButton).toBeEnabled();

        fireEvent.click(decrementButton);

        expect(screen.getByText("0")).toBeInTheDocument();
    });
});
