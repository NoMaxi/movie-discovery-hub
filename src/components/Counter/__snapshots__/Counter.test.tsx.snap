// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Counter Should render the default initial value if not provided in props 1`] = `
<DocumentFragment>
  <div
    class="p-4 flex items-center justify-center gap-4"
  >
    <button
      class="bg-[var(--color-primary)] hover:brightness-80 active:brightness-60 w-[40px] h-[40px] 
                    text-xl font-bold py-2 px-4 rounded-full cursor-pointer
                    opacity-50 cursor-not-allowed"
      disabled=""
    >
      -
    </button>
    <div
      class="flex items-center justify-center text-3xl font-bold min-w-[30px]"
    >
      0
    </div>
    <button
      class="bg-[var(--color-primary)] hover:brightness-80 active:brightness-60 w-[40px] h-[40px]
                    text-xl font-bold py-2 px-4 rounded-full cursor-pointer"
    >
      +
    </button>
  </div>
</DocumentFragment>
`;

exports[`Counter Should render the initial value provided in props 1`] = `
<DocumentFragment>
  <div
    class="p-4 flex items-center justify-center gap-4"
  >
    <button
      class="bg-[var(--color-primary)] hover:brightness-80 active:brightness-60 w-[40px] h-[40px] 
                    text-xl font-bold py-2 px-4 rounded-full cursor-pointer
                    "
    >
      -
    </button>
    <div
      class="flex items-center justify-center text-3xl font-bold min-w-[30px]"
    >
      5
    </div>
    <button
      class="bg-[var(--color-primary)] hover:brightness-80 active:brightness-60 w-[40px] h-[40px]
                    text-xl font-bold py-2 px-4 rounded-full cursor-pointer"
    >
      +
    </button>
  </div>
</DocumentFragment>
`;
